### PR TITLE
Refactor and integration of cerner_splunk_ingredient

### PR DESCRIPTION
**Pull request by alecxvs**
# What? Why?
Integrate cerner_splunk_ingredient. This is a slight refactor and major change. Additional things to note, only Chef 12.14+ and Ruby 2.3+ is supported as of this change. Also contains changes to support search head clustering in newer Splunk versions

# How was it tested?
Unit tests, Vagrant integration testing

### CODE REVIEW

**Comment by nmattam** [comment]
Travis build is failing because of failing rspecs.

File: metadata.rb [hunk 1]
```diff
@@ -1,20 +1,24 @@
...
-depends          'ulimit', '~> 0.3'
-depends          'xml', '~> 1.2'
+depends           'chef-vault', '~> 2.1'
+depends           'xml', '~> 1.2'
+depends           'cerner_splunk_ingredient'
 
-supports         'redhat', '>= 5.5'
-supports         'ubuntu', '>= 12.04'
-supports         'windows', '>= 6.1'
+gem               'unix-crypt'
```
Comment (nmattam) [inline metadata.rb]: What is this for?

**Comment by alecxvs** [comment]
There was a strange issue with old dependencies... namely the xml cookbook, but it's deprecated as of newer Chef versions and I removed it.

Also restored some of the unit tests around sh_cluster. It seems I started to refactor the unit tests, but thought better of it. Or forgot. In any case, that's something that we can deal with later.

File: metadata.rb (hunk 1 above)
Comment (alecxvs) [inline metadata.rb]: Comparing admin password hashes for the search head cluster. Since I introduced reading the admin password from a chef vault, I needed a way to check all of the available passwords (vault, file, and default) to cover all the possible states of the cluster. With this, I can test the passwords against the etc/passwd file before running any commands with it.

File: Vagrantfile [hunk 2]
```diff
@@ -68,8 +70,8 @@ def chef_defaults(chef, name, environment = 'splunk_server')
 end
 
 Vagrant.configure('2') do |config|
-  config.vm.box = 'bento/centos-6.7'
-  config.ohai.primary_nic = 'eth1'
+  config.vm.box = 'bento/centos-7.2'
```
Comment (nmattam) [inline Vagrantfile]: Talked to Alec offline. I'm seeing an error https://gist.github.com/nmattam/4ef676fd38f77f8710e1126e5a333084 and when I revert this change, it works fine.

File: Vagrantfile [hunk 3]
```diff
@@ -52,7 +54,7 @@ def network(config, name, splunk_password = true)
 
   config.berkshelf.enabled = false if Vagrant.has_plugin? 'vagrant-berkshelf'
 
-  config.vm.provision :shell, inline: 'cat /etc/splunk/password; echo' if splunk_password
+  config.vm.provision :shell, inline: "cat /etc/splunk/password || knife vault show cerner_splunk shc_passwords -s 'http://#{@chefip}:4000/' -u #{name} -k /vagrant/vagrant_repo/pems/#{name}.pem; echo" if splunk_password
```
Comment (nmattam) [inline Vagrantfile]: Seeing https://gist.github.com/nmattam/1e1b093616a6cb63863a1da8eebac477 on the search heads.

**Comment by nmattam** [comment]
This might be a Splunk 6.6 issue or me doing something wrong. My updates in the master-apps/_cluster/local/indexes.conf are not getting pushed into the slaves. I'm looking at the slave-apps/_cluster/local/indexes.conf and cannot see the updates. Tried running the apply cluster-bundle command and got this https://gist.github.com/nmattam/f43f5945968e96b3558c47809602e716.

File: spec/unit/recipes/_configure_apps_spec.rb [hunk 4]
```diff
@@ -1,16 +1,21 @@
...
 
 describe 'cerner_splunk::_configure_apps' do
+  let(:chef_run_stubs) {}
+
   subject do
-    runner = ChefSpec::SoloRunner.new(platform: 'centos', version: '6.8') do |node|
-      node.override['splunk']['apps'] = apps
-    end
-    runner.converge('cerner_splunk::_restart_marker', described_recipe)
+    ChefSpec::SoloRunner.new(platform: 'redhat', version: '6.8') do |node|
```
Comment (nmattam) [inline spec/unit/recipes/_configure_apps_spec.rb]: if we are testing on RHEL6.9, we should have it here as well.

File: spec/unit/recipes/_configure_indexes_spec.rb [hunk 5]
```diff
@@ -133,7 +138,10 @@
         }
       }
 
-      expect(subject).to create_splunk_template('system/indexes.conf').with(expected_attributes)
+      expect(subject).to configure_splunk('system/indexes.conf').with(expected_attributes)
+    end
+    it 'should notify the splunk service to restart' do
```
Comment (nmattam) [inline spec/unit/recipes/_configure_indexes_spec.rb]: Maybe also check the notification on apply cluster bundle execute resource.

File: spec/unit/recipes/shc_captain_spec.rb [hunk 6]
```diff
@@ -1,21 +1,31 @@
...
       node.override['splunk']['config']['clusters'] = ['cerner_splunk/cluster']
+      node.override['splunk']['config']['password_secrets'] = { 'shc_captain': 'cerner_splunk/shc_passwords' }
+      node.run_state.merge!('cerner_splunk' => { 'admin_password' => 'changeme' })
     end
     runner.converge(described_recipe)
   end
 
   let(:cluster_config) do
     {
+      'receivers' => ['33.33.33.20'],
```
Comment (nmattam) [inline spec/unit/recipes/shc_captain_spec.rb]: Just curious, do we need to define all these attributes here?

Comment (alecxvs) [inline spec/unit/recipes/shc_captain_spec.rb]: I don't know...

File: spec/unit/recipes/_configure_indexes_spec.rb (hunk 5 above)
Comment (alecxvs) [inline spec/unit/recipes/_configure_indexes_spec.rb]: How important is it to add this test now? I can't actually check for that here since the resource to be notified is in another recipe which does not already have its own spec.

File: spec/unit/recipes/_configure_apps_spec.rb (hunk 4 above)
Comment (alecxvs) [inline spec/unit/recipes/_configure_apps_spec.rb]: f1be08fe47475c2ce52c0c77d26d2eacd92ec8b4

**Comment by nmattam** [comment]
> This might be a Splunk 6.6 issue or me doing something wrong. My updates in the master-apps/_cluster/local/indexes.conf are not getting pushed into the slaves. I'm looking at the slave-apps/_cluster/local/indexes.conf and cannot see the updates. Tried running the apply cluster-bundle command and got this https://gist.github.com/nmattam/f43f5945968e96b3558c47809602e716.

@acharlieh showed me that this was because I didn't wait enough. Although it took some time, the update finally appeared on the slaves.

File: Vagrantfile [hunk 7]
```diff
@@ -287,18 +290,22 @@ Vagrant.configure('2') do |config|
...
+    cfg.omnibus.install_url = "https://packages.chef.io/stable/windows/2012r2/chef-client-#{cfg.omnibus.chef_version}-1-x64.msi"
     cfg.vm.provider :virtualbox do |vb|
       vb.customize ['modifyvm', :id, '--memory', 1024]
+      vb.gui = false
     end
     cfg.vm.provision :chef_client do |chef|
       chef_defaults chef, :f_win2012r2, 'splunk_standalone'
       chef.add_role 'splunk_monitors_windows'
       chef.add_recipe 'cerner_splunk'
+      chef.verbose_logging = true
```
Comment (nmattam) [inline Vagrantfile]: Why turning on verbose logging?

Comment (alecxvs) [inline Vagrantfile]: I was digging deep into the resource update nonsense. It's not necessary (and I don't think it really does anything)

Comment (alecxvs) [inline Vagrantfile]: Actually I'm really not sure why it's in the windows box in particular. That could've been added some time ago. Anyways 9fdb388147a106a667e37d5135d5b9a8b1f5fab7

File: docs/vagrant.md [hunk 8]
```diff
@@ -1,31 +1,33 @@
...
 
 **Note**:
-If you want to set up the vagrant cluster to use the license pools defined in the [license pool hash](databags.md#license-pool-hash) databag, add the `configure_guids` recipe to the run_list on the cluster slave (to update the GUIDs on these slaves to predefined values) and update the `license_uri` attribute in the cluster-vagrant databag item to point to the cluster master (_https://33.33.33.30:8089_).
-After you spin up the cluster slaves you will have to restart the cluster master to bring the cluster to a stable state. While spinning up the cluster slaves, they are re-assigned with different GUIDS by the `configure_guids` recipe which requires a restart (this restart can take a while to complete). Cluster master is restarted so that it can identify the new GUIDS.
+If you want to set up the vagrant cluster to use the license pools defined in the [license pool hash](databags.md#license-pool-hash) databag, add the `configure_guids` recipe to the run*list on the cluster slave (to update the GUIDs on these slaves to predefined values) and update the \`license*uri`attribute in the cluster-vagrant databag item to point to the cluster master (_https://33.33.33.30:8089_).
```
Comment (nmattam) [inline docs/vagrant.md]: formatting seems to be off here.

File: docs/vagrant.md [hunk 9]
```diff
@@ -1,31 +1,33 @@
...
+
+## Spinning up a Search Head Cluster in Vagrant
 
-# Spinning up a Search Head Cluster in Vagrant
+```Bash
+vagrant up /c2\_.\*/
+```
 
-<code>vagrant up /c2_.*/</code>
+c2*boot1 and c2*boot2 are the bootstrap nodes and are provisioned first. c2\_captain is then provisioned and the cluster is established. c2\_deployer is provisioned and pushes apps to the cluster. Finally c2\_newnode is provisioned and joins the cluster as a scale up example.
```
Comment (nmattam) [inline docs/vagrant.md]: "c2\_"

File: docs/vagrant.md (hunk 8 above)
Comment (alecxvs) [inline docs/vagrant.md]: 👀 
8ecb38b96c667de1c89fdb5a659950ff850b2a69

File: docs/vagrant.md (hunk 9 above)
Comment (alecxvs) [inline docs/vagrant.md]: 8ecb38b96c667de1c89fdb5a659950ff850b2a69

File: recipes/_generate_password.rb [hunk 10]
```diff
@@ -8,41 +10,19 @@
...
-  environment 'HOME' => node['splunk']['home']
-  sensitive true
-end
-
-ruby_block 'update admin password in run_state' do
-  block do
-    node.run_state['cerner_splunk']['admin-password'] = new_password
-  end
+# Identify the most specific vault path that matches the current node name and type.
+password_vault_paths = node['splunk']['config']['password_secrets'] || {}
```
Comment (nmattam) [inline recipes/_generate_password.rb]: Did we add this attribute to the readme?

Comment (alecxvs) [inline recipes/_generate_password.rb]: b64f987de12d84caf6c237f12b65bf6c8f46054f

File: libraries/databag.rb [hunk 11]
```diff
@@ -80,12 +82,12 @@ def self.load(string, options = {}) # rubocop:disable CyclomaticComplexity, Perc
       clazz =
         case opts[:type]
         when :simple
-          Chef::DataBagItem
+          Chef::DataBagItem # TODO: ~FC086
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply libraries/databag.rb]: It does fail on any result (`bundle exec foodcritic -f any .`). I wanted to refactor this, thus the TODO, but I think it ended up being way too involved.

File: Gemfile [hunk 12]
```diff
@@ -1,41 +1,11 @@
-source 'https://rubygems.org'
-
-# Going forward, these should be updated to the latest versions immediately post release
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply Gemfile]: 331c7395da05f0115e43dbd3ec50416dfaea0c65

File: Gemfile [hunk 13]
```diff
@@ -1,41 +1,11 @@
...
 gem 'berkshelf'
-gem 'chef', chef_version
-gem 'chef-vault', chef_vault_version
-gem 'chefspec', chefspec_version
-gem 'foodcritic', foodcritic_version
-gem 'rubocop', rubocop_version
+gem 'chef', '= 12.18.31' # https://github.com/cerner/cerner_splunk/issues/142
+gem 'chef-vault'
+gem 'chefspec', '~> 7.1.0'
+gem 'foodcritic', '~> 11.1.0'
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply Gemfile]: 331c7395da05f0115e43dbd3ec50416dfaea0c65

File: libraries/databag.rb [hunk 14]
```diff
@@ -55,17 +57,17 @@ def self.to_value(array, _options = {}) # rubocop:disable CyclomaticComplexity,
       when nil
         nil
       when Array
-        fail "Array '#{array}' can only contain Strings or nil" unless array.all? { |i| i.nil? || i.is_a?(String) }
+        raise "Array '#{array}' can only contain Strings or nil" unless array.all? { |i| i.nil? || i.is_a?(String) }
         data_bag, bag_item, key = array
         Chef::DataBag.validate_name!(data_bag) if data_bag
-        Chef::DataBagItem.validate_id!(bag_item) if bag_item
+        Chef::DataBagItem.validate_id!(bag_item) if bag_item # TODO: ~FC086
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply libraries/databag.rb]: 331c7395da05f0115e43dbd3ec50416dfaea0c65

File: libraries/databag.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (gravesb) [reply libraries/databag.rb]: somehow I missed the todo. yeah, we can refactor this in a separate issue.

File: libraries/authentication.rb [hunk 15]
```diff
@@ -1,134 +1,108 @@
...
+          raise "Unknown type for cacheTiming: #{cacheTiming.class}" unless hash['cacheTiming'].is_a?(Hash)
+          auth_stanzas['cacheTiming'] = hash.delete('cacheTiming')
         end
+
       when 'LDAP'
-        strategies = hash.delete('LDAP_strategies')
-        fail 'LDAP_strategies required for LDAP authentication' unless strategies
-        strategies = [strategies] unless strategies.is_a? Array
+        raise 'LDAP_strategies required for LDAP authentication' unless hash['LDAP_strategies']
+        strategies = [hash.delete('LDAP_strategies')] unless hash['LDAP_strategies'].is_a? Array
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply libraries/authentication.rb]: 6cba4248c93a6cb794f3202c4f87514bed1444c1

File: libraries/authentication.rb [hunk 16]
```diff
@@ -1,134 +1,108 @@
-# coding: UTF-8
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply libraries/authentication.rb]: 6cba4248c93a6cb794f3202c4f87514bed1444c1

File: spec/unit/recipes/_configure_spec.rb [hunk 17]
```diff
@@ -1,14 +1,14 @@
-# coding: UTF-8
+# frozen_string_literal: true
```
Comment (gravesb) [inline spec/unit/recipes/_configure_spec.rb]: There's a warning on line 83 of this file.  You can pull over this line from #155 to resolve it: https://github.com/cerner/cerner_splunk/pull/155/files#diff-2c8e2001d0880b470cbfaef7c90399c4R83

File: recipes/_configure_roles.rb [hunk 18]
```diff
@@ -1,26 +1,26 @@
...
 
-authorize_action = authorize.empty? ? :delete : :create
-
-splunk_template 'system/authorize.conf' do
-  stanzas authorize
-  action authorize_action
-  notifies :touch, 'file[splunk-marker]', :immediately
+splunk_conf 'system/authorize.conf' do
+  config authorize
+  action :configure # Only configure option for now # TODO: What does this mean
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply recipes/_configure_roles.rb]: @acharlieh Do we need to be able to delete this? Should this be overwritten every time instead of 'updated'?

File: resources/splunk_admin_password.rb [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+    raise 'vault_bag and vault_item must both be provided to configure the password vault'
+  end
+end
+
+action_class do
+  def vault
+    return unless vault_bag && vault_item
+    @vault ||= ChefVault::Item.load(vault_bag, vault_item)
+  rescue ChefVault::Exceptions::KeysNotFound, ChefVault::Exceptions::ItemNotFound => e
+    raise e, 'Vault item for admin password does not exist'
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply resources/splunk_admin_password.rb]: 99d91af4e5cb1251a36c06134203de5ee99de3a8

File: recipes/image_prep.rb [hunk 20]
```diff
@@ -1,18 +1,19 @@
...
+#
 # Cookbook Name:: cerner_splunk
 # Recipe:: image_prep
 #
 # Make a universal forwarder part of a host image.
 #
 # Requires Chef 12.6.0 or above
 #
-# See http://docs.splunk.com/Documentation/Forwarder/6.5.2/Forwarder/Makeauniversalforwarderpartofahostimage
+# See http://docs.splunk.com/Documentation/Forwarder/6.6.3/Forwarder/Makeauniversalforwarderpartofahostimage
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply recipes/image_prep.rb]: 99d91af4e5cb1251a36c06134203de5ee99de3a8

File: recipes/_configure_server.rb [hunk 21]
```diff
@@ -121,28 +119,25 @@
...
   case node['splunk']['node_type']
   when :license_server
     'self'
   when :cluster_master, :cluster_slave, :server, :search_head, :shc_search_head, :shc_captain, :shc_deployer
-    if node['splunk']['free_license']
-      'self'
-    else
-      CernerSplunk.my_cluster_data(node)['license_uri'] || 'self'
-    end
+    'self' || !node['splunk']['free_license'] && CernerSplunk.my_cluster_data(node)['license_uri']
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply recipes/_configure_server.rb]: 99d91af4e5cb1251a36c06134203de5ee99de3a8

File: libraries/recipe.rb [hunk 22]
```diff
@@ -64,24 +65,21 @@ def self.keys(node)
...
+    when 'splunk', :splunk then 'splunk'
+    when 'splunkforwarder', :splunkforwarder, :universal_forwarder then 'universal_forwarder'
     end
   end
 
   # Returns filepath to splunk bin based on platform family
   def self.splunk_command(node)
+    # Chef::Log.warn(node['splunk']['home'])
     filepath = "#{node['splunk']['home']}/bin/splunk"
+    # Chef::Log.warn(filepath)
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply libraries/recipe.rb]: 99d91af4e5cb1251a36c06134203de5ee99de3a8

File: spec/unit/recipes/_configure_spec.rb (hunk 17 above)
Comment (alecxvs) [reply spec/unit/recipes/_configure_spec.rb]: I don't have any warning

Comment (gravesb) [reply spec/unit/recipes/_configure_spec.rb]:
```
$ rspec spec/unit/recipes/_configure_spec.rb 

Randomized with seed 5137

cerner_splunk::_configure
  when cluster databag is specified
    includes cerner_splunk::_configure_authentication recipe
    includes cerner_splunk::_configure_inputs recipe
    includes cerner_splunk::_configure_apps recipe
    includes cerner_splunk::_configure_outputs recipe
    includes cerner_splunk::_configure_alerts recipe
    includes cerner_splunk::_configure_server recipe
    includes cerner_splunk::_configure_roles recipe
  when cluster is not specified
    when node is a forwarder
      should raise a warning message
    when node is not a forwarder
WARNING: Using the `raise_error` matcher without providing a specific error or message risks false positives, since `raise_error` will match when Ruby raises a `NoMethodError`, `NameError` or `ArgumentError`, potentially allowing the expectation to pass without even executing the method you are intending to call. Actual error raised was #<RuntimeError: You need to configure at least one cluster databag.>. Instead consider providing a specific error class or message. This message can be suppressed by setting: `RSpec::Expectations.configuration.on_potential_false_positives = :nothing`. Called from /Users/bg010740/.rvm/gems/ruby-2.3.1/gems/chefspec-7.1.0/lib/chefspec/expect_exception.rb:15:in `matches?'.
      should raise an exception
```

File: docs/attributes.md [hunk 23]
```diff
@@ -1,64 +1,60 @@
...
-* `node['splunk']['config']['assumed_index']` - Name of the index to which data is forwarded to by default, when the index is not configured for the input.(`main`)
-* `node['splunk']['bootstrap_shc_member']` - Set this attribute to `true` to bootstrap a member to the Search Head Cluster (SHC). (`false`)
-* `node['splunk']['mgmt_host']` - The host other SHC members use when connecting to the current node. You probably want a wrapper cookbook to override this. (`node['ipaddress']`)
-* `node['splunk']['heavy_forwarder']['use_license_uri']` - Set this attribute to `true` to point the Heavy Forwarder to the license master. (`false`)
-* `node['splunk']['apps']` - An [apps hash](databags.md#apps-hash) of apps to configure locally. (Does not support downloading apps ... yet...)
+## Configurable (with defaults)
 
+- `node['splunk']['exclude_groups']` - Array of local groups to which the splunk user should not belong. Checking existing memberships and removing the splunk user from these groups during the chef run only works on Chef 11.10+, users of prior versions should remove the splunk user from these groups manually. (`['root','wheel','admin']`)
```
Comment (acharlieh) [inline docs/attributes.md]: Chef 11 isn't supported anymore is it?

File: libraries/authentication.rb [hunk 24]
```diff
@@ -1,134 +1,107 @@
...
+    def self.determine_auth_type(settings)
+      config_mapping = {
+        'LDAP_strategies' => 'LDAP',
+        'cacheTiming' => 'Scripted',
+        'scriptPath' => 'Scripted',
+        'scriptSearchFilters' => 'Scripted',
+        'passwordHashAlgorithm' => 'Splunk'
+      }.freeze
+
+      probable_types = config_mapping.values_at(*settings.keys)
```
Comment (acharlieh) [inline libraries/authentication.rb]: This needs a `.uniq` lest you happen to use multiple attributes that indicate the same authType, and the next line blows up.

File: Vagrantfile [hunk 25]
```diff
@@ -30,14 +32,14 @@ end
...
 
-fail 'Non-unique hostnames' if @network.collect { |_, v| v[:hostname] }.uniq!
+raise 'Non-unique hostnames' if @network.collect { |_, v| v[:hostname] }.uniq!
 
-fail 'Non-unique ports' if @network.collect { |_, v| v[:ports].keys }.flat_map { |v| v }.uniq!
+raise 'Non-unique ports' if @network.collect { |_, v| v[:ports].keys }.flat_map { |v| v }.uniq!
 
 def default_omnibus(config)
-  config.omnibus.chef_version = '12'
+  config.omnibus.chef_version = '12.19.36'
```
Comment (acharlieh) [inline Vagrantfile]: But we're pegging to 12.18.31 in the Gemfile ?

Comment (alecxvs) [reply Vagrantfile]: I must have been testing upgrading the chef version but forgot to update the gemfile

Comment (alecxvs) [reply Vagrantfile]: d744f3347c912df0f7e202ef85b2f8935a7b7e07

File: libraries/authentication.rb (hunk 24 above)
Comment (alecxvs) [reply libraries/authentication.rb]: d744f3347c912df0f7e202ef85b2f8935a7b7e07

File: docs/attributes.md (hunk 23 above)
Comment (alecxvs) [reply docs/attributes.md]: d744f3347c912df0f7e202ef85b2f8935a7b7e07

File: recipes/_configure_roles.rb (hunk 18 above)
[the remark this answers is not in the document]
Comment (alecxvs) [reply recipes/_configure_roles.rb]: 6ce3c3ddb9fc3dcbda7450961e403790d1589be1

File: recipes/_configure_shc_roles.rb [hunk 26]
```diff
@@ -15,18 +17,14 @@
...
 
-authorize_action = authorize.empty? ? :delete : :create
-
-splunk_template 'shcluster/_shcluster/authorize.conf' do
-  stanzas authorize
-  action authorize_action
+splunk_conf 'shcluster/apps/_shcluster/authorize.conf' do
+  config authorize
   notifies :run, 'execute[apply-shcluster-bundle]'
+  action :configure
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply recipes/_configure_shc_roles.rb]: 6ce3c3ddb9fc3dcbda7450961e403790d1589be1

File: recipes/_generate_password.rb (hunk 10 above)
Comment (alecxvs) [reply recipes/_generate_password.rb]: 99d91af4e5cb1251a36c06134203de5ee99de3a8

File: spec/unit/recipes/_configure_spec.rb (hunk 17 above)
Comment (alecxvs) [reply spec/unit/recipes/_configure_spec.rb]: aa57872a25700a2dcd3b344f96fcec5afc72f554

File: metadata.rb [hunk 27]
```diff
@@ -1,20 +1,23 @@
...
 
-chef_version     '~> 12.4' if respond_to?(:chef_version)
+source_url        'https://github.com/cerner/cerner_splunk'
+issues_url        'https://github.com/cerner/cerner_splunk/issues'
 
-depends          'chef-vault', '~> 1.3'
-depends          'ulimit', '~> 0.3'
-depends          'xml', '~> 1.2'
+depends           'chef-vault', '~> 3.0'
+depends           'cerner_splunk_ingredient'
```
Comment (gravesb) [inline metadata.rb]: This should probably be pessimistic on major version.  '~> 1.0'

File: metadata.rb (hunk 1 above)
Comment (gravesb) [reply metadata.rb]: We should probably include some kind of version constraint though. Typically a good idea to be pessimistic on the major version.  `'~> 1.0'`

File: metadata.rb (hunk 27 above)
Comment (alecxvs) [reply metadata.rb]: The version is presently 0.1.0, though I suppose I should change that to 1.0 in ingredient too.

Comment (alecxvs) [reply metadata.rb]: ecc3ee815cae26cb5fc0ddb3b51cfb7ea76a20c9

File: metadata.rb (hunk 1 above)
Comment (alecxvs) [reply metadata.rb]: ecc3ee815cae26cb5fc0ddb3b51cfb7ea76a20c9

File: README.md [hunk 28]
```diff
@@ -1,121 +1,129 @@
...
+## Documentation
+
 More in depth documentation including server configuration and data bag formats is located [in this repository](docs/README.md)
 
-License & Authors
------------------
+## License & Authors
+
 - Author:: David Crowder (david.crowder@cerner.com)
 - Author:: Charlie Huggard (charlie.huggard@cerner.com)
```
Comment (acharlieh) [inline README.md]: Add yourself.

File: attributes/_install.rb [hunk 29]
```diff
@@ -10,39 +11,7 @@
     '/etc/splunk'
   end
 
-default['splunk']['package']['version'] = '6.6.0'
-default['splunk']['package']['build'] = '1c4f3bbe1aea'
+default['splunk']['package']['version'] = '6.6.2'
```
Comment (acharlieh) [inline attributes/_install.rb]: Probably should bump to 6.6.3 -> https://www.splunk.com/view/SP-CAAAP3H

File: libraries/app_helpers.rb [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+              owner node['splunk']['user']
+              group node['splunk']['group']
+              mode '0600'
+              content CernerSplunk::ConfigProcs.parse(contents, filename: file_path)
+              action contents.empty? ? :delete : :create
+            end
+          end
+
+          lookups.each do |filename, url|
+            raise "Unsupported lookup file format for #{filename}" unless filename =~ /\.(?:csv\.gz|csv|kmz)$/i
```
Comment (acharlieh) [inline libraries/app_helpers.rb]: Future / unrelated thought: We probably want to log an issue about handling kmz properly. Splunk expands kmz lookups for caching/speed it seems and so manipulating the kmz file we may need to see if we also need to manipulate (delete) the directory too

File: README.md (hunk 28 above)
Comment (alecxvs) [reply README.md]: 31c34317a5b50233361e85cedc7509acb1129192

File: attributes/_install.rb (hunk 29 above)
Comment (alecxvs) [reply attributes/_install.rb]: 31c34317a5b50233361e85cedc7509acb1129192

**Comment by alecxvs** [comment]
Updated to pull cerner_splunk_ingredient from supermarket 🎉 
31c34317a5b50233361e85cedc7509acb1129192

File: metadata.rb [hunk 31]
```diff
@@ -1,20 +1,23 @@
...
-depends          'xml', '~> 1.2'
+depends           'chef-vault', '~> 3.0'
+depends           'cerner_splunk_ingredient', '~> 1.0'
 
-supports         'redhat', '>= 5.5'
-supports         'ubuntu', '>= 12.04'
-supports         'windows', '>= 6.1'
+gem               'unix-crypt', '~> 1.0'
+
+chef_version      '~> 12.16'
```
Comment (wheatevo) [inline metadata.rb]: Would it be possible to instead change this to `chef_version      '>= 12.16'` so that Chef 13 will also be supported? Has this been tested/validated on Chef 13?

File: README.md [hunk 32]
```diff
@@ -1,121 +1,130 @@
...
 [![Code Climate](https://codeclimate.com/github/cerner/cerner_splunk/badges/gpa.svg)](https://codeclimate.com/github/cerner/cerner_splunk)
 
-Requirements
-------------
-* Red Hat Enterprise / CentOS 5.5+ / Windows Server 2008+ (forwarder only) or Ubuntu LTS 12.04+
-* Chef 12+
+## Requirements
+
+- Red Hat Enterprise / CentOS 6.7+ / Windows Server 2008+ (forwarder only) or Ubuntu LTS 12.04+
+- Chef 12.14+
```
Comment (acharlieh) [inline README.md]: nit: 12.14+ or 12.19+ given our tests?

File: docs/contributing.md [hunk 33]
```diff
@@ -1,25 +1,24 @@
...
+## Issue Tracking
+
+- Submit issues in github issues.
+- If it's being tracked internally, you may see references to OPSINFRA and DEVOPS JIRAs
+
+## Communities & Resources
+
+- [Splunk Enterprise Documentation](http://docs.splunk.com/Documentation/Splunk)
+- [Splunk Answers (Q&A)](http://answers.splunk.com/)
+- [Splunk IRC](http://www.splunk.com/view/SP-CAAACDF) #splunk on [EFNet](http://www.efnet.org/)
```
Comment (acharlieh) [inline docs/contributing.md]: Nit as we're reformatting, we could add:
`- [Splunk Slack](https://splunk-usergroups.slack.com) Request an invite [here](https://splunk-usergroups.signup.team)`

File: libraries/app_helpers.rb [hunk 34]
```diff
@@ -0,0 +1,90 @@
...
+            file_path = app_path + 'lookups' + filename
+
+            directory file_path.to_s do
```
Comment (acharlieh) [inline libraries/app_helpers.rb]: This probably needs to be `file_path.dirname.to_s` ?

File: libraries/app_helpers.rb [hunk 35]
```diff
@@ -0,0 +1,90 @@
...
+              owner node['splunk']['user']
+              group node['splunk']['group']
+              mode '0755'
+            end
+
+            remote_file file_path.to_s do
+              if (url ||= '').empty?
+                action :delete
+              else
```
Comment (acharlieh) [inline libraries/app_helpers.rb]: Probably need to be passing the source parameter in this else

File: libraries/app_helpers.rb [hunk 36]
```diff
@@ -0,0 +1,90 @@
...
+        files = (params[:files] || {}).select { |key, _| key.to_s =~ /.+(?<!\.conf)$/ }
+        lookups = params[:lookups] || {}
+
+        proc do |app_base|
+          app_path = Pathname.new app_base
+
+          files.each do |filename, contents|
+            file_path = app_path + filename
+
+            directory file_path.to_s do
```
Comment (acharlieh) [inline libraries/app_helpers.rb]: likewise, this should be file_path.dirname.to_s

File: libraries/app_helpers.rb [hunk 37]
```diff
@@ -0,0 +1,90 @@
...
+              group node['splunk']['group']
+              mode '0755'
+              recursive true
+            end
+
+            file file_path.to_s do
+              owner node['splunk']['user']
+              group node['splunk']['group']
+              mode '0600'
+              content CernerSplunk::ConfigProcs.parse(contents, filename: file_path)
```
Comment (acharlieh) [inline libraries/app_helpers.rb]: I think this is more applicable for conf files below than up here?

File: libraries/authentication.rb [hunk 38]
```diff
@@ -1,134 +1,107 @@
...
+    def self.determine_auth_type(settings)
+      config_mapping = {
+        'LDAP_strategies' => 'LDAP',
+        'cacheTiming' => 'Scripted',
+        'scriptPath' => 'Scripted',
+        'scriptSearchFilters' => 'Scripted',
+        'passwordHashAlgorithm' => 'Splunk'
+      }.freeze
+
+      probable_types = config_mapping.values_at(*settings.keys).uniq
```
Comment (acharlieh) [inline libraries/authentication.rb]: If settings['authType'] is set, you want to include this in the array that you then uniq over.

File: libraries/authentication.rb [hunk 39]
```diff
@@ -1,134 +1,107 @@
...
+        script['scriptSearchFilters'] = hash.delete 'scriptSearchFilters' if hash.key? 'scriptSearchFilters'
         auth_stanzas['script'] = script
-        cache_timing = hash.delete('cacheTiming')
-        if cache_timing
-          fail "Unknown type for cacheTiming: #{cacheTiming.class}" unless cache_timing.is_a?(Hash)
-          auth_stanzas['cacheTiming'] = cache_timing
+        hash['authSettings'] = 'script'
+
+        if hash.key? 'cacheTiming'
+          raise "Unknown type for cacheTiming: #{cacheTiming.class}" unless hash['cacheTiming'].is_a?(Hash)
```
Comment (acharlieh) [inline libraries/authentication.rb]: The variable cacheTiming from the message doesn't exist.  (it's an existing issue obviously)

File: libraries/config_procs.rb [hunk 40]
```diff
@@ -0,0 +1,70 @@
...
+      def self.constant(value:, **_)
+        proc { value }
+      end
+
+      def self.vault(coordinate:, default_coords: nil, pick_context: nil, **_)
+        proc { CernerSplunk::DataBag.load coordinate, type: :vault, default: default_coords, pick_context: pick_context }
+      end
+
+      def self.existing(filename:, **_)
+        proc do |context, config|
```
Comment (acharlieh) [inline libraries/config_procs.rb]: Hm... we seem to have changed the API for this proc, Note to self: review changes to application of procs.

**Comment by gravesb** [comment]
This PR is far behind master and most of the developers/reviewers on it are no longer working on this project. Closing since I don't think we'll ever get back to this project.
